backends/ubpf/ubpfProgram.cpp

```diff
@@ -0,0 +1,303 @@
+            ::error("Expected toplevel package %1% to have 3 parameters", pack->type);
+            return false;
+        }
+
+        auto pb = pack->getParameterValue(model.filter.parser.name)
+                ->to<IR::ParserBlock>();
+        BUG_CHECK(pb != nullptr, "No parser block found");
+        parser = new UBPFParser(this, pb, typeMap);
+        success = parser->build();
+        if (!success)
+            return success;
+
+        auto cb = pack->getParameterValue(model.filter.control.name)
+                ->to<IR::ControlBlock>();
+        BUG_CHECK(cb != nullptr, "No control block found");
+        control = new UBPFControl(this, cb, parser->headers);
+        success = control->build();
+
+        if (!success)
+            return success;
+
+        auto dpb = pack->getParameterValue(model.filter.deparser.name)
+                ->to<IR::ControlBlock>();
+        BUG_CHECK(dpb != nullptr, "No deparser block found");
+        deparser = new UBPFDeparser(this, dpb, parser->headers);
+        success = deparser->build();
+
+        return success;
+    }
+
+    void UBPFProgram::emitC(UbpfCodeBuilder *builder, cstring headerFile) {
+        emitGeneratedComment(builder);
+
+        builder->appendFormat("#include \"%s\"", headerFile);
+        builder->newline();
+
+        builder->target->emitIncludes(builder);
+        emitPreamble(builder);
+        emitUbpfHelpers(builder);
+
+        builder->emitIndent();
+        builder->target->emitMain(builder, "entry", contextVar.c_str());
+        builder->blockStart();
+
+        emitPktVariable(builder);
+
+        emitHeaderInstances(builder);
+        builder->append(" = ");
+        parser->headerType->emitInitializer(builder);
+        builder->endOfStatement(true);
+
+        emitMetadataInstance(builder);
+        builder->append(" = ");
+        parser->metadataType->emitInitializer(builder);
+        builder->endOfStatement(true);
+
+        emitLocalVariables(builder);
+        emitPacketCheck(builder);
+        builder->newline();
+        builder->emitIndent();
+        builder->appendFormat("goto %s;", IR::ParserState::start.c_str());
+        builder->newline();
+
+        parser->emit(builder);
+
+        emitPipeline(builder);
+
+        deparser->emit(builder);
+
+        builder->emitIndent();
+        builder->appendFormat("if (%s)\n", control->passVariable);
+        builder->increaseIndent();
+        builder->emitIndent();
+        builder->appendFormat("return %s;\n", builder->target->forwardReturnCode().c_str());
+        builder->decreaseIndent();
+        builder->emitIndent();
+        builder->appendLine("else");
+        builder->increaseIndent();
+        builder->emitIndent();
+        builder->appendFormat("return %s;\n", builder->target->dropReturnCode().c_str());
+        builder->decreaseIndent();
+        builder->blockEnd(true);
+    }
+
+    void UBPFProgram::emitUbpfHelpers(EBPF::CodeBuilder *builder) const {
+        builder->append(
```

Comment:
Shouldn't these be in some method of UBPFTarget?

Reply:
Moved to UBPFTarget.